### ISSUE TITLE
What was the downside to `eval-when-compile`ing dir constants?

### ISSUE THREAD

**Issue by ivanbrennan** [hlissner/doom-emacs#4769]
Doom used to leverage `eval-when-compile` when declaring its directory constants, e.g.:
```elisp
(defconst doom-emacs-dir
  (eval-when-compile (expand-file-name user-emacs-directory)))

(defconst doom-core-dir
  (eval-when-compile (expand-file-name "core" doom-emacs-dir)))
```
but at a certain point (906c6e205), it stopped doing so:
```elisp
(defconst doom-emacs-dir
  (expand-file-name user-emacs-directory))

(defconst doom-core-dir
  (expand-file-name "core" doom-emacs-dir))
```

I'm curious to know what motivated this change. I searched the git log and PRs/Issues, but came up empty.

**Comment by hlissner** [comment]
I did it to offload work onto the byte-compiler. Things in `eval-when-compile` forms are evaluated at compile time and replaced with its return value. I thought it could shave a couple ms from repeating `expand-file-name` calls, which may be expensive depending on what's in `file-name-handler-alist`. But, I was just throwing stuff at the wall to see what stuck, and the benchmarks turned up wanting.

The gains were negligible and byte-compiling proved problematic to support, especially for a codebase as active as Doom's, further exacerbated when people other than myself began using Doom. Now, I just use string concatenation to avoid that perceived cost (though, we're still talking about negligible margins; especially so that Doom unsets `file-name-handler-alist` at startup). 

TL;DR A premature optimization came and went. Not much to see here.

----

Side note: please use [our Discord](https://doomemacs.org/discord) or [our Discourse](https://discourse.doomemacs.org) (will be up soon) for posts like this. Our issue tracker is for bugs and actionable tickets.

**Comment by ivanbrennan** [comment]
@hlissner Thanks for the detailed explanation!
And apologies for the noise -- I'll use Discord/Discourse for future questions.
